internals/osutil/mkdir.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2014-2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package osutil
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+
+	"github.com/canonical/pebble/internals/osutil/sys"
+)
+
+// XXX: we need to come back and fix this; this is a hack to unblock us.
+// Have a lock so that if one goroutine tries to mkdirallchown /foo/bar, and
+// another tries to mkdirallchown /foo/baz, they can't both decide they need
+// to make /foo and then have one fail.
+var mu sync.Mutex
+
+// MkdirOptions holds the options for a call to Mkdir.
+type MkdirOptions struct {
+	// If false (the default), return an error if the parent directory is missing.
+	// If true, any missing parent directories of the given path are created
+	// (with the same permissions and owner/group as the leaf directory).
+	MakeParents bool
+
+	// If false (the default), return an error if the target directory already exists.
+	// If true, don't return an error if the target directory already exists (unless
+	// it's not a directory).
+	ExistOK bool
+
+	// If false (the default), no explicit chown is performed.
+	// If true, perform an explicit chown on any directories created, using the UserID
+	// and GroupID provided.
+	Chown bool
+
```

Comment:
Let's remove the blank lines between `Chown` and `UserID` and `GroupID` so they're clearly grouped together.